### PR TITLE
Streaming schema backup tool

### PR DESCRIPTION
**Pull request by jjaakola-aiven**
# About this change - What it does

Streaming backup file restore and export support.

# Why this way

The backup export and Avro anonymized schema export created one large array of content before writing to file. For large schema registries this can consume a lot of memory. The data can be streamed to file single schema message at a time from the schema topic.

For restoring the former format is discarded and per line Base16 encoded file is expected for input. First line in the file is the version header `/V2`.
The restore can read the schema message line by line instead of reading and parsing the whole exported JSON array.

# Memory profiling results

Results are from fairly large dataset (~370 K messages) and collected with Memray. Elapsed time is from Linux `time` command.

## Restore
Elapsed time for restore could be optimized by sending data in batches to schema topic. Current and previous implementation are similar and send one message at a time. After calling `producer.send` a call to `flush` is done immediately. Also the returned future from `send` is waited before progressing to next message.

Earlier version:
```
Elapsed time: 5:03.21
Total number of allocations: 21554637
Total number of frames seen: 2987
Peak memory usage: 3.3 GB
```
JSON lines version:
```
Elapsed time: 5:18.93
Total number of allocations: 25141916
Total number of frames seen: 3008
Peak memory usage: 15.2 MB
```

Base16 version:
```
Elapsed time: 4:50.64
Total number of allocations: 24746793
Total number of frames seen: 2997
Peak memory usage: 15.2 MB
```
Removal of `flush` and `get` after each restored message and sending to partition 0 drops the restore time significantly.
Sending to partition 0 has a reduction of ~10 seconds, this skips the partitioner (murmur2).
```
Elapsed: 1:25.21
```

## Export
Earlier version:
```
Elapsed time: 0:11.98
Total number of allocations: 6232591
Total number of frames seen: 2939
Peak memory usage: 1.8 GB
```
JSON lines version:
```
Elapsed: 0:11.03
Total number of allocations: 7721438
Total number of frames seen: 2938
Peak memory usage: 22.0 MB
```

Base16 version:
```
Elapsed time: 0:07.95
Total number of allocations: 4446491
Total number of frames seen: 2933
Peak memory usage: 21.9 MB
```

### CODE REVIEW

File: karapace/schema_backup.py [hunk 1]
```diff
@@ -24,10 +25,27 @@
 LOG = logging.getLogger(__name__)
 
 
+JsonData = Any
```
Comment (hackaugusto) [inline karapace/schema_backup.py:28]: Please use the typing:

https://github.com/aiven/karapace/blob/eb5a7720475b66de5011722db4799e3815fdacaa/karapace/typing.py#L3

Comment (jjaakola-aiven) [reply karapace/schema_backup.py]: Done.

File: karapace/schema_backup.py [hunk 2]
```diff
@@ -158,81 +163,75 @@ def restore_backup(self):
...
+                        ser = export_func(key=key, value=value)
+                        if ser:
+                            fp.write(ser)
+                            fp.write("\n")
 
-        return values
+        if self.backup_location:
+            LOG.info("Anonymized Avro schema export written to %r", self.backup_location)
+        else:
+            LOG.info("Anonymized Avro schema export written to stdout")
```
[the remark this answers is not in the document]
Comment (jjaakola-aiven) [reply karapace/schema_backup.py]: Changed to `Schema export written to %r`

File: karapace/schema_backup.py [hunk 3]
```diff
@@ -158,81 +163,75 @@ def restore_backup(self):
...
+                self._restore_backup(fp)
+            else:
+                self._restore_backup_version_1_single_array(fp)
+        self.close()
+
+    def _handle_restore_message(self, item: Tuple[str, str]):
+        key = encode_value(item[0])
+        value = encode_value(item[1])
+        future = self.producer.send(self.topic_name, key=key, value=value)
+        self.producer.flush(timeout=self.timeout_ms)
```
[the remark this answers is not in the document]
Comment (jjaakola-aiven) [reply karapace/schema_backup.py]: I don't these are needed. Schemas topic is a special topic with one partition and the order is guaranteed.
In the end the `close` is called and that will flush.

**Review by tvainika — COMMENTED** [review]
Overall looks good. However now there is a new v2 format for backups, which is tailored to Karapace needs I'd like to see header line defining format version instead of heuristic (hex versus json opening bracket) to detect format version. That would allow later upgrades to format if ever needed.

Second I think it would  make sense add now backup restore tests that have fixed v1 format generated and v2 generated backup files, where restore code is tested to avoid breaking any restore code.

**Comment by tvainika** [comment]
> Second I think it would make sense add now backup restore tests that have fixed v1 format generated and v2 generated backup files, where restore code is tested to avoid breaking any restore code.

@jjaakola-aiven I now see there are tests for this, thanks for clarifying test names. What I actually had in my mind was that having v1 & v2 backup files as separate artefact files on the repository would make those better distinguishable. Currently there is only test code generating temporary files for restore test. Having stable artefact file works both as an example of backup file, and also separates format of the backup file from the test logic.

**Review by tvainika — APPROVED** [review]

**Pull request merged**